File: Source/AlphaTab.JavaScript/Platform/JavaScript/AlphaSynthWebWorker.cs
using System;
using AlphaTab.Audio.Synth;
using AlphaTab.Audio.Synth.Midi;
using AlphaTab.Audio.Synth.Synthesis;
using AlphaTab.Haxe;
using AlphaTab.Haxe.Js;
using AlphaTab.Haxe.Js.Html;
using AlphaTab.Model;
using AlphaTab.Util;
using Phase;

namespace AlphaTab.Platform.JavaScript
{
    /// <summary>
    /// This class implements a HTML5 WebWorker based version of alphaSynth
    /// which can be controlled via WebWorker messages.
    /// </summary>
    class AlphaSynthWebWorker
    {
        #region Commands

        public const string CmdPrefix = "alphaSynth.";

        // Main -> Worker
        public const string CmdInitialize = CmdPrefix + "initialize";

        public const string CmdSetLogLevel = CmdPrefix + "setLogLevel";
        public const string CmdSetMasterVolume = CmdPrefix + "setMasterVolume";
        public const string CmdSetMetronomeVolume = CmdPrefix + "setMetronomeVolume";
        public const string CmdSetPlaybackSpeed = CmdPrefix + "setPlaybackSpeed";
        public const string CmdSetTickPosition = CmdPrefix + "setTickPosition";
        public const string CmdSetTimePosition = CmdPrefix + "setTimePosition";
        public const string CmdSetPlaybackRange = CmdPrefix + "setPlaybackRange";
        public const string CmdSetIsLooping = CmdPrefix + "setIsLooping";

        public const string CmdPlay = CmdPrefix + "play";
        public const string CmdPause = CmdPrefix + "pause";
        public const string CmdPlayPause = CmdPrefix + "playPause";
        public const string CmdStop = CmdPrefix + "stop";
        public const string CmdLoadSoundFontBytes = CmdPrefix + "loadSoundFontBytes";
        public const string CmdLoadMidi = CmdPrefix + "loadMidi";
        public const string CmdSetChannelMute = CmdPrefix + "setChannelMute";
        public const string CmdSetChannelSolo = CmdPrefix + "setChannelSolo";
        public const string CmdSetChannelVolume = CmdPrefix + "setChannelVolume";
        public const string CmdSetChannelProgram = CmdPrefix + "setChannelProgram";
        public const string CmdResetChannelStates = CmdPrefix + "resetChannelStates";

        // Worker -> Main
        public const string CmdReady = CmdPrefix + "ready";
        public const string CmdReadyForPlayback = CmdPrefix + "readyForPlayback";
        public const string CmdPositionChanged = CmdPrefix + "positionChanged";
        public const string CmdPlayerStateChanged = CmdPrefix + "playerStateChanged";
        public const string CmdFinished = CmdPrefix + "finished";
        public const string CmdSoundFontLoaded = CmdPrefix + "soundFontLoaded";
        public const string CmdSoundFontLoadFailed = CmdPrefix + "soundFontLoadFailed";
        public const string CmdMidiLoaded = CmdPrefix + "midiLoaded";
        public const string CmdMidiLoadFailed = CmdPrefix + "midiLoadFailed";
        public const string CmdLog = CmdPrefix + "log";

        #endregion

        private readonly Audio.Synth.AlphaSynth _player;
        private readonly DedicatedWorkerGlobalScope _main;

        public AlphaSynthWebWorker(DedicatedWorkerGlobalScope main, string id)
        {
            _main = main;
            _main.AddEventListener("message", (Action<MessageEvent>)HandleMessage);

            _player = new Audio.Synth.AlphaSynth(new AlphaSynthWorkerSynthOutput());
            _player.PositionChanged += OnPositionChanged;
            _player.PlayerStateChanged += OnPlayerStateChanged;
            _player.Finished += OnFinished;
            _player.SoundFontLoaded += OnSoundFontLoaded;
            _player.SoundFontLoadFailed += OnSoundFontLoadFailed;
            _player.SoundFontLoadFailed += OnSoundFontLoadFailed;
            _player.MidiLoaded += OnMidiLoaded;
            _player.MidiLoadFailed += OnMidiLoadFailed;
            _player.ReadyForPlayback += OnReadyForPlayback;

            _main.PostMessage(new { cmd = CmdReady });
        }

        public static void Init()
        {
            DedicatedWorkerGlobalScope main = Lib.Global;
            main.AddEventListener("message", (Action<MessageEvent>)(e =>
            {
                var data = e.Data;
                string cmd = data.cmd;
                switch (cmd)
                {
                    case CmdInitialize:
                        AlphaSynthWorkerSynthOutput.PreferredSampleRate = data.sampleRate;
                        Logger.LogLevel = data.logLevel;
                        new AlphaSynthWebWorker(main, data.id);
                        break;
                }
            }));
        }

        public void HandleMessage(MessageEvent e)
        {
            var data = e.Data;
            string cmd = data.cmd;
            switch (cmd)
            {
                case CmdSetLogLevel:
                    Logger.LogLevel = data.value;
                    break;
                case CmdSetMasterVolume:
                    _player.MasterVolume = data.value;
                    break;
                case CmdSetMetronomeVolume:
                    _player.MetronomeVolume = data.value;
                    break;
                case CmdSetPlaybackSpeed:
                    _player.PlaybackSpeed = data.value;
                    break;
                case CmdSetTickPosition:
                    _player.TickPosition = data.value;
                    break;
                case CmdSetTimePosition:
                    _player.TimePosition = data.value;
                    break;
                case CmdSetPlaybackRange:
                    _player.PlaybackRange = data.value;
                    break;
                case CmdSetIsLooping:
                    _player.IsLooping = data.value;
                    break;
                case CmdPlay:
                    _player.Play();
                    break;
                case CmdPause:
                    _player.Pause();
                    break;
                case CmdPlayPause:
                    _player.PlayPause();
                    break;
                case CmdStop:
                    _player.Stop();
                    break;
                case CmdLoadSoundFontBytes:
                    _player.LoadSoundFont(data.data);
                    break;
                case CmdLoadMidi:
                    _player.LoadMidi(JsonConverter.JsObjectToMidiFile(data.midi));
                    break;
                case CmdSetChannelMute:
                    _player.SetChannelMute(data.channel, data.mute);
                    break;
                case CmdSetChannelSolo:
                    _player.SetChannelSolo(data.channel, data.solo);
                    break;
                case CmdSetChannelVolume:
                    _player.SetChannelVolume(data.channel, data.volume);
                    break;
                case CmdSetChannelProgram:
                    _player.SetChannelProgram(data.channel, data.program);
                    break;
                case CmdResetChannelStates:
                    _player.ResetChannelStates();
                    break;
            }
        }


        public void OnPositionChanged(PositionChangedEventArgs e)
        {
            _main.PostMessage(new
            {
                cmd = CmdPositionChanged,
                currentTime = e.CurrentTime,
                endTime = e.EndTime,
                currentTick = e.CurrentTick,
                endTick = e.EndTick
            });
        }

        public void OnPlayerStateChanged(PlayerStateChangedEventArgs e)
        {
            _main.PostMessage(new
            {
                cmd = CmdPlayerStateChanged,
                state = e.State
            });
        }

        public void OnFinished(bool isLooping)
        {
            _main.PostMessage(new
            {
                cmd = CmdFinished,
                isLooping = isLooping
            });
        }

        public void OnSoundFontLoaded()
        {
            _main.PostMessage(new
            {
                cmd = CmdSoundFontLoaded
            });
        }

        public void OnSoundFontLoadFailed(Exception e)
        {
            _main.PostMessage(new
            {
                cmd = CmdSoundFontLoadFailed,
                error = SerializeException(e)
            });
        }

        private object SerializeException(Exception e)
        {
            dynamic error = Json.Parse(Json.Stringify(e));
            dynamic e2 = e;
            if (e2.message)
            {
                error.message = e2.message;
            }
            if (e2.stack)
            {
                error.stack = e2.stack;
            }
            if (e2.constructor && e2.constructor.name)
            {
                error.type = e2.constructor.name;
            }

            return error;
        }

        public void OnMidiLoaded()
        {
            _main.PostMessage(new
            {
                cmd = CmdMidiLoaded
            });
        }

        public void OnMidiLoadFailed(Exception e)
        {
            _main.PostMessage(new
            {
                cmd = CmdMidiLoaded,
                error = SerializeException(e)
            });
        }

        public void OnReadyForPlayback()
        {
            _main.PostMessage(new
            {
                cmd = CmdReadyForPlayback
            });
        }

        public void SendLog(LogLevel level, string s)
        {
            _main.PostMessage(new
            {
                cmd = CmdLog,
                level = level,
                message = s
            });
        }
    }
}

File: src/platform/javascript/AlphaSynthWebWorker.ts
import { AlphaSynth, type IAlphaSynthAudioExporter } from '@src/synth/AlphaSynth';
import type { PlayerStateChangedEventArgs } from '@src/synth/PlayerStateChangedEventArgs';
import type { PositionChangedEventArgs } from '@src/synth/PositionChangedEventArgs';
import { JsonConverter } from '@src/model/JsonConverter';
import { AlphaSynthWorkerSynthOutput } from '@src/platform/javascript/AlphaSynthWorkerSynthOutput';
import type { IWorkerScope } from '@src/platform/javascript/IWorkerScope';
import { Logger } from '@src/Logger';
import { Environment } from '@src/Environment';
import type { MidiEventsPlayedEventArgs } from '@src/synth/MidiEventsPlayedEventArgs';
import type { PlaybackRangeChangedEventArgs } from '@src/synth/PlaybackRangeChangedEventArgs';

/**
 * This class implements a HTML5 WebWorker based version of alphaSynth
 * which can be controlled via WebWorker messages.
 * @target web
 */
export class AlphaSynthWebWorker {
    private _player: AlphaSynth;
    private _main: IWorkerScope;
    private _exporter: Map<number, IAlphaSynthAudioExporter> = new Map<number, IAlphaSynthAudioExporter>();

    public constructor(main: IWorkerScope, bufferTimeInMilliseconds: number) {
        this._main = main;
        this._main.addEventListener('message', this.handleMessage.bind(this));

        this._player = new AlphaSynth(new AlphaSynthWorkerSynthOutput(), bufferTimeInMilliseconds);
        this._player.positionChanged.on(this.onPositionChanged.bind(this));
        this._player.stateChanged.on(this.onPlayerStateChanged.bind(this));
        this._player.finished.on(this.onFinished.bind(this));
        this._player.soundFontLoaded.on(this.onSoundFontLoaded.bind(this));
        this._player.soundFontLoadFailed.on(this.onSoundFontLoadFailed.bind(this));
        this._player.soundFontLoadFailed.on(this.onSoundFontLoadFailed.bind(this));
        this._player.midiLoaded.on(this.onMidiLoaded.bind(this));
        this._player.midiLoadFailed.on(this.onMidiLoadFailed.bind(this));
        this._player.readyForPlayback.on(this.onReadyForPlayback.bind(this));
        this._player.midiEventsPlayed.on(this.onMidiEventsPlayed.bind(this));
        this._player.playbackRangeChanged.on(this.onPlaybackRangeChanged.bind(this));
        this._main.postMessage({
            cmd: 'alphaSynth.ready'
        });
    }

    public static init(): void {
        const main: IWorkerScope = Environment.globalThis as IWorkerScope;
        main.addEventListener('message', e => {
            const data: any = e.data;
            const cmd: string = data.cmd;
            switch (cmd) {
                case 'alphaSynth.initialize':
                    AlphaSynthWorkerSynthOutput.preferredSampleRate = data.sampleRate;
                    Logger.logLevel = data.logLevel;
                    Environment.globalThis.alphaSynthWebWorker = new AlphaSynthWebWorker(
                        main,
                        data.bufferTimeInMilliseconds
                    );
                    break;
            }
        });
    }

    public handleMessage(e: MessageEvent): void {
        const data: any = e.data;
        const cmd: string = data.cmd;
        switch (cmd) {
            case 'alphaSynth.setLogLevel':
                Logger.logLevel = data.value;
                break;
            case 'alphaSynth.setMasterVolume':
                this._player.masterVolume = data.value;
                break;
            case 'alphaSynth.setMetronomeVolume':
                this._player.metronomeVolume = data.value;
                break;
            case 'alphaSynth.setPlaybackSpeed':
                this._player.playbackSpeed = data.value;
                break;
            case 'alphaSynth.setTickPosition':
                this._player.tickPosition = data.value;
                break;
            case 'alphaSynth.setTimePosition':
                this._player.timePosition = data.value;
                break;
            case 'alphaSynth.setPlaybackRange':
                this._player.playbackRange = data.value;
                break;
            case 'alphaSynth.setIsLooping':
                this._player.isLooping = data.value;
                break;
            case 'alphaSynth.setCountInVolume':
                this._player.countInVolume = data.value;
                break;
            case 'alphaSynth.setMidiEventsPlayedFilter':
                this._player.midiEventsPlayedFilter = data.value;
                break;
            case 'alphaSynth.play':
                this._player.play();
                break;
            case 'alphaSynth.pause':
                this._player.pause();
                break;
            case 'alphaSynth.playPause':
                this._player.playPause();
                break;
            case 'alphaSynth.stop':
                this._player.stop();
                break;
            case 'alphaSynth.playOneTimeMidiFile':
                this._player.playOneTimeMidiFile(JsonConverter.jsObjectToMidiFile(data.midi));
                break;
            case 'alphaSynth.loadSoundFontBytes':
                this._player.loadSoundFont(data.data, data.append);
                break;
            case 'alphaSynth.resetSoundFonts':
                this._player.resetSoundFonts();
                break;
            case 'alphaSynth.loadMidi':
                this._player.loadMidiFile(JsonConverter.jsObjectToMidiFile(data.midi));
                break;
            case 'alphaSynth.setChannelMute':
                this._player.setChannelMute(data.channel, data.mute);
                break;
            case 'alphaSynth.setChannelTranspositionPitch':
                this._player.setChannelTranspositionPitch(data.channel, data.semitones);
                break;
            case 'alphaSynth.setChannelSolo':
                this._player.setChannelSolo(data.channel, data.solo);
                break;
            case 'alphaSynth.setChannelVolume':
                this._player.setChannelVolume(data.channel, data.volume);
                break;
            case 'alphaSynth.resetChannelStates':
                this._player.resetChannelStates();
                break;
            case 'alphaSynth.destroy':
                this._player.destroy();
                this._main.postMessage({
                    cmd: 'alphaSynth.destroyed'
                });
                break;
            case 'alphaSynth.applyTranspositionPitches':
                this._player.applyTranspositionPitches(new Map<number, number>(JSON.parse(data.transpositionPitches)));
                break;
        }

        if (cmd.startsWith('alphaSynth.exporter')) {
            this.handleExporterMessage(e);
        }
    }
    private handleExporterMessage(e: MessageEvent) {
        const data: any = e.data;
        const cmd: string = data.cmd;
        try {
            switch (cmd) {
                case 'alphaSynth.exporter.initialize':
                    const exporter = this._player.exportAudio(
                        data.options,
                        JsonConverter.jsObjectToMidiFile(data.midi),
                        data.syncPoints,
                        data.transpositionPitches
                    );
                    this._exporter.set(data.exporterId, exporter);

                    this._main.postMessage({
                        cmd: 'alphaSynth.exporter.initialized',
                        exporterId: data.exporterId
                    });
                    break;

                case 'alphaSynth.exporter.render':
                    if (this._exporter.has(data.exporterId)) {
                        const exporter = this._exporter.get(data.exporterId)!;
                        const chunk = exporter.render(data.milliseconds);
                        this._main.postMessage({
                            cmd: 'alphaSynth.exporter.rendered',
                            exporterId: data.exporterId,
                            chunk
                        });
                    } else {
                        this._main.postMessage({
                            cmd: 'alphaSynth.exporter.error',
                            exporterId: data.exporterId,
                            error: new Error('Unknown exporter ID')
                        });
                    }
                    break;

                case 'alphaSynth.exporter.destroy':
                    this._exporter.delete(data.exporterId);
                    break;
            }
        } catch (e) {
            this._main.postMessage({
                cmd: 'alphaSynth.exporter.error',
                exporterId: data.exporterId,
                error: e
            });
        }
    }

    public onPositionChanged(e: PositionChangedEventArgs): void {
        this._main.postMessage({
            cmd: 'alphaSynth.positionChanged',
            currentTime: e.currentTime,
            endTime: e.endTime,
            currentTick: e.currentTick,
            endTick: e.endTick,
            isSeek: e.isSeek,
            originalTempo: e.originalTempo,
            modifiedTempo: e.modifiedTempo
        });
    }

    public onPlayerStateChanged(e: PlayerStateChangedEventArgs): void {
        this._main.postMessage({
            cmd: 'alphaSynth.playerStateChanged',
            state: e.state,
            stopped: e.stopped
        });
    }

    public onFinished(): void {
        this._main.postMessage({
            cmd: 'alphaSynth.finished'
        });
    }

    public onSoundFontLoaded(): void {
        this._main.postMessage({
            cmd: 'alphaSynth.soundFontLoaded'
        });
    }

    public onSoundFontLoadFailed(e: any): void {
        this._main.postMessage({
            cmd: 'alphaSynth.soundFontLoadFailed',
            error: this.serializeException(Environment.prepareForPostMessage(e))
        });
    }

    private serializeException(e: any): unknown {
        const error: any = JSON.parse(JSON.stringify(e));
        if (e.message) {
            error.message = e.message;
        }
        if (e.stack) {
            error.stack = e.stack;
        }
        if (e.constructor && e.constructor.name) {
            error.type = e.constructor.name;
        }
        return error;
    }

    public onMidiLoaded(e: PositionChangedEventArgs): void {
        this._main.postMessage({
            cmd: 'alphaSynth.midiLoaded',
            currentTime: e.currentTime,
            endTime: e.endTime,
            currentTick: e.currentTick,
            endTick: e.endTick,
            isSeek: e.isSeek,
            originalTempo: e.originalTempo,
            modifiedTempo: e.modifiedTempo
        });
    }

    public onMidiLoadFailed(e: any): void {
        this._main.postMessage({
            cmd: 'alphaSynth.midiLoaded',
            error: this.serializeException(Environment.prepareForPostMessage(e))
        });
    }

    public onReadyForPlayback(): void {
        this._main.postMessage({
            cmd: 'alphaSynth.readyForPlayback'
        });
    }

    public onMidiEventsPlayed(args: MidiEventsPlayedEventArgs): void {
        this._main.postMessage({
            cmd: 'alphaSynth.midiEventsPlayed',
            events: args.events.map(JsonConverter.midiEventToJsObject)
        });
    }

    public onPlaybackRangeChanged(args: PlaybackRangeChangedEventArgs): void {
        this._main.postMessage({
            cmd: 'alphaSynth.playbackRangeChanged',
            playbackRange: args.playbackRange
        });
    }
}
